pkgs/by-name/bi/bisq/package.nix

```diff
@@ -0,0 +1,157 @@
+{
+  stdenvNoCC,
+  lib,
+  makeWrapper,
+  fetchurl,
+  makeDesktopItem,
+  copyDesktopItems,
+  imagemagick,
+  openjdk,
+  dpkg,
+  writeScript,
+  bash,
+  tor,
+  zip,
+  gnupg,
+}:
+
+let
+  bisq-launcher =
+    args:
+    writeScript "bisq-launcher" ''
+      #! ${bash}/bin/bash
+
+      # This is just a comment to convince Nix that Tor is a
+      # runtime dependency; The Tor binary is in a *.jar file,
+      # whereas Nix only scans for hashes in uncompressed text.
+      # ${bisq-tor}
+
+      rm -fR $HOME/.local/share/Bisq2/tor
+
+      exec "${openjdk}/bin/java" -Djpackage.app-version=@version@ -classpath @out@/lib/app/desktop-app-launcher.jar:@out@/lib/app/* ${args} bisq.desktop_app_launcher.DesktopAppLauncher "$@"
+    '';
+
+  bisq-tor = writeScript "bisq-tor" ''
+    #! ${bash}/bin/bash
+
+    exec ${tor}/bin/tor "$@"
+  '';
+
+  # Alejandro Garcia's public key was obtained from
+  # https://bisq.network/pubkey/E222AA02.asc
+  publicKey = ./pubkey.asc;
+in
+stdenvNoCC.mkDerivation rec {
+  pname = "bisq2";
+  version = "2.1.2";
+
+  src = fetchurl {
+    url = "https://github.com/bisq-network/${pname}/releases/download/v${version}/Bisq-${version}.deb";
+    sha256 = "0zgv70xlz3c9mrwmiaa1dgagbc441ppk2vrkgard8zjrvk8rg7va";
+
+    # Verify the upstream Debian package prior to extraction.
+    # See https://bisq.wiki/Downloading_and_installing#Verify_installer_file
+    # This ensures that a successful build of this Nix package requires the Debian
```

Comment:
This is a good idea except that when the key expires it will break and be no longer reproducible.

Reply:
When I was planning out this package it was my intention to make it reproducible. However, I had not considered the impact of an expiring key.

I performed a test to see what happens when an expired key is used to verify a signature:

```
❯ gpg --list-keys
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
/home/emmanuel/projects/gpgtest/./gnupg/pubring.kbx
---------------------------------------------------
pub   ed25519 2024-10-28 [SC] [expired: 2024-10-29]
      8A9DE59F1481A17B052F4746C325905B2798233C
uid           [ expired] test <test@test.com>

❯ gpg --verify you-didnt-come-this-far.jpg.sig
gpg: assuming signed data in 'you-didnt-come-this-far.jpg'
gpg: Signature made Mon 28 Oct 2024 04:02:14 PM EDT
gpg:                using EDDSA key 8A9DE59F1481A17B052F4746C325905B2798233C
gpg: Good signature from "test <test@test.com>" [expired]
gpg: Note: This key has expired!
Primary key fingerprint: 8A9D E59F 1481 A17B 052F  4746 C325 905B 2798 233C

~/projects/gpgtest
❯ echo $?
0
```

The test shows that an expired key does not affect verification.

Now that upstream started adding the keys to the Github release assets, I modified the package to use those. That should also help ensure the builds are reproducible.

Reply:
Does it factor in that the signing was done before the key expired?

Reply:
As far as I can tell, there's no way to ensure that a signature was created prior to the key expiring.

GPG does refuse to create a signature using an expired key, but certainly that refusal can be circumvented.

This reveals a weakness in the verification process since there's no way to tell if upstream's deb package was signed with an expired key or not. If I were to modify the package to fail upon detecting an expired key, then it would prevent old versions of the package from building.

Since I have the two public keys in my key ring, I can manually check the expiration dates when updating this package, as a form of mitigation.